### PR TITLE
introduce Containerd-snapshotter feature flag

### PR DESCRIPTION
**Pull request by ndeloof**
**- What I did**
introduce Containerd-snapshotter feature flag

**- How I did it**
added support for `Containerd-snapshotter`feature flag
introduced `usesSnapshotter` for usability in code
illustration in oci_linux for some obvious places where moby enforce a mounted rootFS, while migration to containerd will rely on an active snapshot

**- How to verify it**
no impact as this is just basic skaffolding to bootstrap containerd image store migration work. Enabling this new feature flag is expected to just break until some concrete development takes place.

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**
image

### CODE REVIEW

File: daemon/oci_linux.go
Comment (thaJeztah) [inline daemon/oci_linux.go:731]: I see there's various locations where `s.Root` is referenced; should this be initialised with an empty `specs.Root` if `daemon.usesSnapshotter()` (to prevent panics), e.g. this may panic? https://github.com/moby/moby/blob/9ccedde9703bbef70881a0e8a3006310b7d3d788/libcontainerd/remote/client.go#L131


Or should the field be propagated in some other way? There's also the ReadOnly option that is discarded now; is that passed on through some other ways?

Comment (ndeloof) [reply daemon/oci_linux.go:731]: enabling this flag in the current codebase **will break**, this PR just introduces the feature flag and utility method, with an initial illustration example where it is obviously needed, but it is not expected enabling this flag won't trigger a panic.

File: daemon/daemon.go
Comment (ndeloof) [reply daemon/daemon.go:152]: good point. Doing so would be unpredictable as existing containers will be either set with a snapshot or moby-managed rootfs, and won't be manageable by the code once flag has changed value. Same applies anyway to any flags mechanism that will require daemon to be restarted.

Maybe a feature flag is not the best option then. I'd be fine we decide to rely on a build flag if this makes more sense.

Comment (thaJeztah) [reply daemon/daemon.go:152]: Thinking what's the best option as well. Basically, we have a (somewhat) similar situation if you switch to a different storage-driver, but for that case, the daemon is able to detect the previous selection; perhaps we need something similar for this (detect what the daemon was configured with?) 🤔

**Comment by AkihiroSuda** [comment]
Thanks for working on this, but not sure this is ready to merge until it actually works (for docker pull at least)

**Comment by ndeloof** [comment]
@AkihiroSuda our goal is to avoid a log term branch with massive changes, and doing so we will propose small incremental changes which are by nature incomplete, and won't be able to offer a working daemon. So this feature flag is a way for us to get maintainer reviews, but enforce this won't impact Moby stability on master
See https://github.com/rumpl/moby/tree/containerd-store for initial exploratory work that will drive this effort

**Review by rumpl — APPROVED** [review]

**Review by khs1994 — APPROVED** [review]

**Review by tianon — APPROVED** [review]
LGTM, but IMO we should cut a 22.06 branch before we merge it so it doesn't sneak into 22.06 yet :innocent:

**Review by AkihiroSuda — APPROVED** [review]

**Pull request merged**
